dbms/src/Interpreters/ExpressionActions.cpp

```diff
@@ -329,16 +382,51 @@ void ExpressionAction::prepare(Block & sample_block, const Settings & settings)
 }
 
 
-void ExpressionAction::execute(Block & block, bool dry_run) const
+template <bool execute_on_block>
+void ExpressionAction::execute(
+    Block & block,
```

Comment:
Putting every other line in a function under an `#ifdef` (oh sorry it's called `if constexpr` now) is a time-tested, respectable way to make it totally unreadable. If you have to perform the same actions on two kinds of data structures, you could make them support the same interface, either by changing them or by writing a wrapper.